File: Data Structures & Algorithms/is-palindrome/submission-3.java
class Solution {
    public boolean isPalindrome(String s) {
        
        int l = 0, r = s.length()-1;
        char[] arr = s.toLowerCase().toCharArray();

        while (l < r) {
            while (!isAlphaNumeric(arr[l]) && l < r) {
                l++;
            }
            while (!isAlphaNumeric(arr[r]) && l < r) {
                r--;
            }
            if (arr[l] != arr[r]) {
                return false;
            }
            l++;
            r--;
        }
        return true;
    }

    private boolean isAlphaNumeric(char c) {
        return (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9');
    }
}

File: Data Structures & Algorithms/valid-palindrome-ii/submission-3.ts
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    validPalindrome(s: string): boolean {

        let l = 0, r = s.length - 1;

        while (l < r) {
            if (s[l] == s[r]) {
                l++;
                r--;
            }
            // console.log(s[l] + " " + s[r])
            // if (r - l == 1) {
            //     return true
            // }
            else {
                return this.isPalindrome(l, r - 1, s) || this.isPalindrome(l+1, r, s)
            }
        }
        return true
    }

    isPalindrome(l: number, r: number, s: string): boolean {
        while (l <= r) {
            if (s[l] != s[r]) {
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
}
